3D transform utilities

KrisBraun opened the pull request:
Based on #10982 by @brandonmcconnell.

Add 3D transform utilities. All are applied to specific CSS properties rather than as `transform` functions.

| Utility | CSS Property | CSS Value |
|--------|--------|--------|
| `rotate` | `rotate` | arbitrary values can be used to define an axis of rotation<br>(e.g. `rotate_[1_2_3_45deg]`) |
| `rotate-x` | `rotate` | same as `rotate`, applied in the x-dimension; see note below |
| `rotate-y` | `rotate` | same as `rotate`, applied in the y-dimension; see note below |
| `scale` | `scale` | arbitrary values are now supported<br>(e.g. `scale-[2_1.5_3]`, `scale-[var(--custom-scale)]`) |
| `scale-z` | `scale` | same as `scale`, applied only to the z-dimension |
| `scale-3d` | `scale` | applies `scale` across the z-dimension<br>(e.g. `scale-50 scale-3d` is equivalent to `scale-50 scale-z-50`) |
| `translate-z` | `translate` | same as `translate`, applied only to the z-dimension |
| `translate-3d` | `translate` | applies `translate` across the z-dimension<br>(e.g. `translate-full translate-3d` is equivalent to `translate-full translate-z-full`) |
| `perspective` |  `perspective` |  `dramatic` (`100px`), `near` (`300px`), `normal` (`500px`), `midrange` (`800px`), `distant` (`1200px`), or arbitrary (e.g. `perspective-[42px]`) |
| `perspective-origin` | `perspective-origin` | same as `origin` (`transform-origin`) |
| `transform` | `transform` | arbitrary values (e.g. `transform-[rotateX(45deg)_rotateY(-90deg)]`) are now supported |
| `transform-flat` | `transform-style` | `flat` |
| `transform-3d` | `transform-style` | `preserve-3d` |
| `transform-content` | `transform-box` | `content-box` |
| `transform-border` | `transform-box` | `border-box` |
| `transform-fill` | `transform-box` | `fill-box` |
| `transform-stroke` | `transform-box` | `stroke-box` |
| `transform-view` | `transform-box` | `view-box` |
| `backface-visible` | `backface-visibility` | `visible` |
| `backface-hidden` | `backface-visibility` | `hidden` |

*Note on `rotate`, `rotate-x`, and `rotate-y`*: Rotations are applied using the `rotate` property, which supports a single angle of rotation along with an axis of rotation. `rotate` defaults to the z-dimension and can specify other dimensions (including a vector) using an arbitrary value. Applying multiple rotations in a sequence is order-dependent and can be done using an arbitrary `transform` value (e.g. `transform-[rotateX(45deg)_rotateY(-90deg)]`). 

Care has been taken to only apply transformations in the z-dimension when the relevant utilities are used. This should avoid triggering GPU rendering except when needed.

KrisBraun committed:
3D rotation utilities

RobinMalfait commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1222,21 +1222,51 @@ export function createUtilities(theme: Theme) {
+        return [decl('rotate', `${value} var(--tw-rotate)`)]
+      }
+    } else {
+      value = theme.resolve(candidate.value.value, ['--rotate']) ?? `${candidate.value.value}deg`
```
We should validate if `candidate.value.value` in the fallback case `${candidate.value.value}deg`. Similar to how we handle it in this pr: #13245 

Otherwise `rotate-foo` will turn into:
```css
.rotate-foo {
  rotate: foodeg;
}
```

KrisBraun replied:
`rotate-potato` test added for safety. 🥔

RobinMalfait commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1222,21 +1222,51 @@ export function createUtilities(theme: Theme) {
   ])
+  // rotate-x, rotate-y, and rotate-z change the axis of rotation from the default z-axis.
+  // Arbitrary vector values are implemented in the functional variant.
+  ;['x', 'y', 'z'].forEach((axis) => {
```
For consistency, I think we should write this as:

```ts
for (let axis of ['x', 'y', 'z']) {
  // ...
}
```

RobinMalfait reviewed (commented):
I know this is still in draft mode, but saw it and wanted to comment already before I forgot.

KrisBraun committed:
Validate rotate values

KrisBraun committed:
Replace forEach with for loop

KrisBraun committed:
transform-style, transform-box, and backface-visibility utilities

KrisBraun committed:
Tests for transform utilities

KrisBraun committed:
'perspective' utility

KrisBraun committed:
Fix tests

thecrypticace commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1384,6 +1420,20 @@ export function createUtilities(theme: Theme) {
+    {
+      valueThemeKeys: ['--perspective'],
+    },
+  ])
```
You should be able to drop this `suggest` line since we don't support bare values or modifiers.

KrisBraun replied:
Ah, gotcha. I thought this was required for the theme keys, but I see `functionalUtility` handles that.

thecrypticace commented on packages/tailwindcss/src/utils/infer-data-type.ts:
```diff
@@ -283,3 +287,38 @@ function isBackgroundSize(value: string) {
+ * <units> = deg | rad | grad | turn
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/CSS/angle
+ */
```
this comment should be attached to the function — for consistency with some of the others and so it shows up in documentation popups / hovers.

thecrypticace commented on packages/tailwindcss/src/utils/infer-data-type.ts:
```diff
@@ -283,3 +287,38 @@ function isBackgroundSize(value: string) {
+
+const IS_VECTOR = new RegExp(`^${HAS_NUMBER.source} +${HAS_NUMBER.source} +${HAS_NUMBER.source}$`)
+
+function isVector(value: string) {
```
Ditto here too.

KrisBraun committed:
Remove unnecessary suggestion; move function comments

KrisBraun committed:
scale-z

KrisBraun committed:
Fix Intellisense test

KrisBraun committed:
perspective-origin

KrisBraun committed:
scale-3d

KrisBraun committed:
Only include the z component of scale if it's defined

We want to avoid triggerring unnecessary 3D transformations.

KrisBraun committed:
Comment the reason for setting --tw-rotate

KrisBraun committed:
Test full bare rotate

KrisBraun committed:
Fix merge

KrisBraun committed:
Comment on rotate arbitrary value

adamwathan commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1381,70 +1441,74 @@ export function createUtilities(theme: Theme) {
    * @css `scale`
    */
-  functionalUtility('scale-y', {
+  functionalUtility('scale-3d', {
```
I wonder if there's a way to make this a static utility that acts like a flag instead of a new suite of classes? APIs like `scale-3d-150` feel kinda yuck to me.

If we could do `scale-150 scale-3d` I think that might feel nicer, might have to think through what `scale-x-150 scale-3d` should mean but I think it feels solveable.

KrisBraun replied:
Yes, that's much better! [Implemented here](https://github.com/tailwindlabs/tailwindcss/pull/13248/commits/03a5c47dd8fa423e6eb4f919c540a38511c276c9).

Effectively, `scale` sets the scale in all three dimensions, but only applies it in two. `scale-3d` applies across all three. `scale-x`, `scale-y`, and `scale-z` can be used to set (or override) the scale in individual dimensions.

In the case of `scale-x-150 scale-3d`, the `scale-3d` has no effect, as the scale in the z-dimension is `1`.

KrisBraun replied:
`scale-z` also applies in three dimensions, so `scale-3d` is not required alongside `scale-z`.

brandonmcconnell replied:
Could we still support a way for 3d transforms to be made in a single utility, like `scale3d-[2,1.5,3]`

Especially for something that gets transitioned as often as transform properties, it would be useful to be able to interpret these values onto element and reference them via a single var.

```html
<div
  style={{
    '--scale3d': scale3d
  }}
  class="scale3d-[--scale3d]"
/>
```

Without this, we could still interpret values this way, but we would have to break each axis for each transformation into its own CSS variable and connected tail and utility.

```html
<div
  style={{
    '--scale-x': scaleX,
    '--scale-y': scaleY,
    '--scale-z': scaleZ,
  }}
  class="scale-x-[--scale-x] scale-y-[--scale-y] scale-z-[--scale-z] scale-3d"
/>
```

I think, as long as the below still works, we can just use more complex 3d situations like this when needed:

```html
<div
  style={{
    '--scale3d': scale3d
  }}
  class="[scale3d:var(--scale3d)]"
/>
```

KrisBraun replied:
Yes, an escape hatch for arbitrary values is helpful. I've added that so your first example can be written `scale-[2_1.5_3]` and the last can be simplified to `scale-[var(--scale3d)]`.

brandonmcconnell replied:
Thanks! I presume `scale-[--scale3d]` would also work (without `var()`)?

KrisBraun replied:
Yes, it will support whatever Tailwind CSS v4 supports in general. Bare variables lead to a number of parsing ambiguities, so there's a chance support for them could be deprecated/dropped, but no decisions have been made yet.

adamwathan commented on packages/tailwindcss/theme.css:
```diff
@@ -421,6 +421,13 @@
+  --perspective-near: 300px;
+  --perspective-normal: 500px;
+  --perspective-midrange: 800px;
+  --perspective-distant: 1200px;
```
Makes sense that we need to add these to the theme if we want a named scale but also a bummer because we worked so hard to cut back on the size of the default theme. Anyone have thoughts on just making this purely numeric/bare-value based? I'm like 75% for named, 25% for numeric but curious what others think. Admittedly this is a really good example of a property where people would benefit for a small curated set of options, because nobody knows what the hell these numbers mean.

thecrypticace replied:
I like the named scale as well. It's conceptually distance from the screen so bare values seems fine to me too but I think a small, default set of values is more beneficial.

brandonmcconnell replied:
Based on our discussion in https://github.com/tailwindlabs/tailwindcss/pull/10982, I also think this small set of named values would be very helpful here, plus allowing for numeric in case someone wants to do `perspective-[3000]` (or `perspective-3000` if you're dropping brackets as part of v4).

KrisBraun replied:
Thanks, @brandonmcconnell, that helped me spot that we weren't supporting bare values (e.g. `perspective-123`), just arbitrary values (e.g. `perspective-[123px]`). Both are now supported.

adamwathan commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1261,19 +1275,47 @@ export function createUtilities(theme: Theme) {
+      // Vector components for the axis of rotation
+      property('--tw-rotate-x', '0', '<number>'),
+      property('--tw-rotate-y', '0', '<number>'),
+      property('--tw-rotate-z', '0', '<number>'),
```
In my head I was thinking we'd just have two variables:

```
property('--tw-rotate', '0deg', '<angle>'),
property('--tw-rotate-axis'),
```

Then the values for `--tw-rotate-axis` would be something like:

`[x | y | z | <number{3}]`

...however you do that with `syntax` in `@property`.

This means you couldn't do this:

```
<div class="rotate-45 rotate-x rotate-y">
```

...but I think that doesn't bother me personally, you'd just have to use a vector for that:

```
<div class="rotate-45 rotate-[1_1_0]">
```

I'm not sure how common it actually is to want to rotate something by exactly the same angle on two different axis — my gut is probably not common?

With this change we wouldn't have to translate `rotate-x` to a vector, it could just set `--tw-rotate-axis: x` because `rotate: x 45deg` is valid.

thecrypticace replied:
To add to this — we could (later) add `rotate-xy` as a utility too. Which seems just as explanatory as two separate utilities. So I'm Adam on this one. A single property for the axis feels like a good way to do this.

brandonmcconnell replied:
Would we still be able to do something like `rotate-x-45 rotate-y-135 rotate-z-20`?

Having granular control over individual axis transforms in a pragmatic way rather than using a vector is an important feature here IMO.

This is possible, and can even be converted into a vector under the hood if need be, which I demo'd here:
https://x.com/branmcconnell/status/1768009776938844550

KrisBraun replied:
We've further simplified this by specifying the rotation axis via a modifier on `rotate`:

| Class | CSS |
| ---- | ---- |
| `rotate-45` | `rotate: 45deg` |
| `rotate-45/x` | `rotate: x 45deg` |
| `rotate-45/y` | `rotate: y 45deg` |
| `rotate-45/z` | `rotate: z 45deg` |
| `rotate-45/[1_2_3]` | `rotate: 1 2 3 45deg` |

This makes it clear angles are not composable, in line with what CSS supports via the [`rotate` property](https://developer.mozilla.org/en-US/docs/Web/CSS/rotate).

Composing multiple angles requires a pipeline using `transform`. That is now supported via arbitrary `transform` values (e.g. `transform-[rotateX(45deg) rotateY(-90deg)]`).

brandonmcconnell replied:
@KrisBraun Yes, I was primarily speaking about using multi-axis rotation using the traditional syntax instead of the vector syntax.

This is possible using even the regular `rotate` properties if you convert them to vector syntax under the hood. I demonstrate this in the twitter thread I linked above ([also here](https://x.com/branmcconnell/status/1768009776938844550)).

So hypothetically, someone could do this and get the expected behavior:
```
rotate-45/x rotate-30/y rotate-60/z
```

KrisBraun replied:
After more thought, we decided `rotate-x` and `rotate-y` are more in line with other Tailwind utilities, even if they are mutually exclusive. The PR and description have been updated.

KrisBraun replied:
@brandonmcconnell Totally understand the appeal of composable rotation angles. They're achievable using an arbitrary `transform` value as shown above. It's more verbose but has at least a couple advantages:

1. The order of rotations is explicitly specified.
2. It's clear the `transform` property is being modified, rather than `rotate`.

There are also a few reasons not to translate multiple rotations into a vector:

1. CSS variables can't be supported if the translation is done in code at compile time.
2. Implementing the translation with CSS math functions is technically possible, but ridiculously long and likely slow.
3. Any translation would need to assume an order of composition, unless that was also configurable, ballooning the complexity.
4. The translated values would be harder to understand in dev tools.

We're going with the grain of CSS on this. The `rotate` property is clearly designed for applying a single angle of rotation, and `transform` is the escape hatch for more complex transformation pipelines.

Thanks so much for initiating this PR! We're now at a point where the simple things are simple and everything else is possible. We're looking forward to seeing what folks build with this!

RobinMalfait commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1381,70 +1441,74 @@ export function createUtilities(theme: Theme) {
-      if (Number.isNaN(Number(value))) return null
-      return `${value}%`
-    },
+    handleBareValue: ({ value }) => `${value}%`,
```
I think something went wrong while rebasing, this should have some validation:
```suggestion
    handleBareValue: ({ value }) => {
       if (Number.isNaN(Number(value))) return null
       return `${value}%`
     },
```

Otherwise, we can type `scale-3d-potato%`

MoARABY commented:
it will be useful property

KrisBraun committed:
perspective bare values

Support `perspective-123` (but not `perspective-potato`)

KrisBraun committed:
scale-3d as a static modifier to scale

Instead of scale-3d taking a separate scale, it modifies scale to apply in three dimensions.

KrisBraun committed:
Test that scale-x overrides scale

KrisBraun committed:
scale arbitrary values

Support arbitrary value for scale (e.g. `scale-[1_2_3.5]`).

KrisBraun committed:
Specify rotation axis using a modifier

Support single rotation angles in line with the [CSS `rotate` property](https://developer.mozilla.org/en-US/docs/Web/CSS/rotate). Using modifiers (e.g. `rotate-45/x`) makes it clearer that the axis of rotation is modified. Thanks @adamwathan for this suggestion.

Composing angles is only supported in CSS via a pipeline of `transform` functions. I'll add arbitrary value support to `transform` next as an escape hatch for those cases that need more complex transformations.

KrisBraun committed:
Use property defaults for scale-3d

KrisBraun committed:
`transform` arbitrary values

Support arbitrary values for `transform`. The `skew-x` and `skew-y` transforms are applied before any arbitrary transformations.

KrisBraun committed:
Add translate-z and translate-3d

Both work the same way as scale-z and scale-3d.

KrisBraun committed:
Add translate-[xyz]-px

KrisBraun commented:
`translate-z` and `translate-3d` have been added. They work similarly to `scale` and `scale-3d`. (Also, missing `translate-[xyz]-px` utilities were added.)

KrisBraun committed:
Comment on how skewX and skewY are applied

KrisBraun committed:
Remove unnecessary suggest

KrisBraun committed:
Simplify translate

KrisBraun committed:
Fix up comment on rotate syntax

KrisBraun committed:
Back to rotate-x and rotate-y rather than rotate modifiers

RobinMalfait commented on packages/tailwindcss/src/utilities.ts:
```diff
@@ -1282,6 +1309,27 @@ export function createUtilities(theme: Theme) {
+  for (let axis of ['x', 'y']) {
+    functionalUtility(`rotate-${axis}`, {
+      supportsNegative: true,
+      supportsFractions: true,
```
I don't think we are doing anything with the fraction, so I think we can drop this?

RobinMalfait commented on packages/@tailwindcss-postcss/src/__snapshots__/index.test.ts.snap:
```diff
@@ -393,6 +393,11 @@ exports[`\`@import 'tailwindcss'\` is replaced with the generated CSS 1`] = `
     --line-height-8: 2rem;
     --line-height-9: 2.25rem;
     --line-height-10: 2.5rem;
+    --perspective-dramatic: 100px;
+    --perspective-near: 300px;
+    --perspective-normal: 500px;
+    --perspective-midrange: 800px;
+    --perspective-distant: 1200px;
```
Are these the final names for the default values we ship?

KrisBraun replied:
Yes, taken from [here](https://github.com/tailwindlabs/tailwindcss/pull/10982#issuecomment-1710739315).

RobinMalfait reviewed (commented):
Looking good, a few questions remaining

KrisBraun committed:
3D transform test fixes
